### PR TITLE
Implement Optimistic Effects

### PR DESCRIPTION
**Pull request by charlesBochet**
I'm introducing a new pattern called `optimistic effects`.

Optimistic Effect is a way to declare that query result cache should be updated when some mutations are performed and native apollo cache is failing to update them. 

## Problem

The typical use case is the following:

Let's imagine a query: `GetPeopleQuery({ where: { name { like { '%toto%'} }})`

Apollo does not know whenever it needs to add or remove rows from the query cache whenever
- a person is renamed to 'toto' for example (this person should be added to the query cache
- a person is renamed from 'toto' to something else
- a person is created with the name 'toto'
- (apollo does handle the person deletion using cache.evict())

## Proposed Solution

When performing a query, you can registerOptimisticEffect() in the query and attach it to a typename. This optimistic effect can precisely describe what to do when an entity is added or modified (kind of mimicking the backend behavior). Usually, you want to read the existing query cache, and update it.

Ex: OptimisticEffect Declaration:
```
function optimisticEffectResolver({
  cache,
  entities,
  variables,
}: {
  cache: ApolloCache<Company>;
  entities: Company[];
  variables: GetCompaniesQueryVariables;
}) {
  const existingData = cache.readQuery<GetCompaniesQuery>({
    query: GET_COMPANIES,
    variables: { orderBy: variables.orderBy, where: variables.where },
  });

  if (!existingData) {
    return;
  }

  cache.writeQuery({
    query: GET_COMPANIES,
    variables: { orderBy: variables.orderBy, where: variables.where },
    data: {
      people: [...entities, ...existingData.companies],
    },
  });
}

export function getCompaniesOptimisticEffect(
  variables: GetCompaniesQueryVariables,
) {
  return {
    key: 'generic-entity-table-data-companies',
    variables: variables,
    typename: 'Company',
    resolver: optimisticEffectResolver,
  };
}
```

Register:
```
useGetRequest({
    variables: { orderBy, where: whereFilters },
    onCompleted: (data: any) => {
      const entities = data[getRequestResultKey] ?? [];
      setEntityTableData(entities, filterDefinitionArray);
      registerOptimisticEffect(
        getCompaniesOptimisticEffect({ orderBy, where: whereFilters }),
      );
    },
  });
```

Once registered, an optimistic effect is kept in memory.

Then, while performing a mutation, you can call triggerOptimisticEffects(typename) and all registered effect tied to this typename will be triggered

Example:
`triggerOptimisticEffects('Company', [data?.createOneCompany]);`

### CODE REVIEW

File: front/src/modules/apollo/optimistic-effect/hooks/useOptimisticEffect.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        set(optimisticEffectState, {
+          ...optimisticEffects,
+          [key]: optimisticEffect,
+        });
+      },
+  );
+
+  const triggerOptimisticEffects = useRecoilCallback(
+    ({ snapshot }) =>
+      (typename: string, entities: any[]) => {
```
Comment (lucasbordeau) [inline front/src/modules/apollo/optimistic-effect/hooks/useOptimisticEffect.ts:27]: Would be nice to derivate a string literal from somewhere, but it's ok as a string

File: front/src/modules/companies/graphql/optimistic-effects/getCompaniesOptimisticEffect.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (!existingData) {
+    return;
+  }
+
+  cache.writeQuery({
+    query: GET_COMPANIES,
+    variables: { orderBy: variables.orderBy, where: variables.where },
+    data: {
+      people: [...entities, ...existingData.companies],
```
Comment (lucasbordeau) [inline front/src/modules/companies/graphql/optimistic-effects/getCompaniesOptimisticEffect.ts:33]: `companies:` ?

File: front/src/modules/apollo/optimistic-effect/states/optimisticEffectState.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { atom } from 'recoil';
+
+import { OptimisticEffect } from '../types/OptimisticEffect';
+
+export const optimisticEffectState = atom<
+  Record<string, OptimisticEffect<unknown, unknown>>
+>({
+  key: 'optimisticEffectState',
+  default: {},
```
Comment (magrinj) [inline front/src/modules/apollo/optimistic-effect/states/optimisticEffectState.ts:9]: @charlesBochet Don't we need to keep optimistic effect state in local storage, as Apollo cache is stored inside it ?

File: front/src/modules/apollo/optimistic-effect/hooks/useOptimisticEffect.ts (hunk 1 above)
Comment (charlesBochet) [reply front/src/modules/apollo/optimistic-effect/hooks/useOptimisticEffect.ts:27]: I think we won't be able as we move to the flexible schema anyway!

File: front/src/modules/companies/graphql/optimistic-effects/getCompaniesOptimisticEffect.ts (hunk 2 above)
Comment (charlesBochet) [reply front/src/modules/companies/graphql/optimistic-effects/getCompaniesOptimisticEffect.ts:33]: yes, good catch!

File: front/src/modules/apollo/optimistic-effect/states/optimisticEffectState.ts (hunk 3 above)
Comment (charlesBochet) [reply front/src/modules/apollo/optimistic-effect/states/optimisticEffectState.ts:9]: I've found no official way to access the list of queries through apollo API, so I prefer to maintain my own state, rather that to take a dependency on something that won't be stable

**Review by magrinj — APPROVED** [review]

**Pull request merged**